pkg/services/publicdashboards/api/middleware.go

```diff
@@ -2,33 +2,47 @@ package api
 
 import (
 	"net/http"
-	"strconv"
 
 	"github.com/grafana/grafana/pkg/infra/metrics"
 	"github.com/grafana/grafana/pkg/models"
 	"github.com/grafana/grafana/pkg/services/publicdashboards"
+	"github.com/grafana/grafana/pkg/services/publicdashboards/internal/tokens"
 	"github.com/grafana/grafana/pkg/web"
 )
 
-func SetPublicDashboardFlag() func(c *models.ReqContext) {
+// Adds orgId to context based on org of public dashboard
+func SetPublicDashboardOrgIdOnContext(publicDashboardService publicdashboards.Service) func(c *models.ReqContext) {
 	return func(c *models.ReqContext) {
-		// TODO: Find a better place to set this, or rename this function
-		orgIDValue := c.Req.URL.Query().Get("orgId")
-		orgID, err := strconv.ParseInt(orgIDValue, 10, 64)
-		if err == nil && orgID > 0 && orgID != c.OrgID {
-			c.OrgID = orgID
+		// Check access token is present on the request
+		accessToken, _ := web.Params(c.Req)[":accessToken"]
+		if accessToken == "" {
+			return
+		}
+
+		// Get public dashboard
+		pd, _, err := publicDashboardService.GetPublicDashboard(c.Req.Context(), accessToken)
+		if err != nil || pd == nil {
+			return
+		}
+
+		if pd.IsEnabled {
+			c.OrgID = pd.OrgId
 		}
-		c.IsPublicDashboardView = true
 	}
 }
 
+// Adds public dashboard flag on context
+func SetPublicDashboardFlag(c *models.ReqContext) {
+	c.IsPublicDashboardView = true
+}
+
 func RequiresValidAccessToken(publicDashboardService publicdashboards.Service) func(c *models.ReqContext) {
 	return func(c *models.ReqContext) {
 		accessToken, ok := web.Params(c.Req)[":accessToken"]
 
 		// Check access token is present on the request
-		if !ok || accessToken == "" {
-			c.JsonApiErr(http.StatusBadRequest, "Invalid access token", nil)
+		if !ok || !tokens.IsValidAccessToken(accessToken) {
```

Comment:
If the request is not valid because the `accessToken` is malformed we should return 400.
If it is valid but there is no dashboard with that access token we should return 404.

Reply:
Done